### ISSUE TITLE
Picture of EBM after a bit of training

### ISSUE THREAD

**Issue by josephmaa** [ldoshi/rome-wasnt-built-in-a-day#245]
<img width="712" height="776" alt="Image" src="https://github.com/user-attachments/assets/184642a9-c85a-40f3-99d7-77886d27062c" />

**Comment by josephmaa** [comment]
Testing histogram visualization bucketed by k number of bits, each bucket is in lexicographic order. To be used for toy example of energy based model

<img width="1552" height="676" alt="Image" src="https://github.com/user-attachments/assets/29506060-0e7f-47c6-ad6e-2e4f5b5165f9" />

**Comment by josephmaa** [comment]
Testing visualization of ebm model as it progresses through training that I had at the end of the day, something messed up with visualization:

<img width="1552" height="676" alt="Image" src="https://github.com/user-attachments/assets/381bf8b7-6c2e-48ef-a8c4-255784669f4f" />

Code for reference:
```
import torch
import torch.nn as nn
import torch.optim as optim
import numpy as np
import matplotlib.pyplot as plt
from itertools import combinations, product

# --------- EBM model ---------
class MLP(nn.Module):
    def __init__(self, input_size, hidden_sizes):
        super().__init__()
        layers = []
        prev = input_size
        for h in hidden_sizes:
            layers.append(nn.Linear(prev, h))
            layers.append(nn.ReLU())
            prev = h
        layers.append(nn.Linear(prev, 1))  # output energy
        self.network = nn.Sequential(*layers)
        
    def forward(self, x):
        return self.network(x)

# --------- Hyperparameters ---------
n = 4
hidden_sizes = [32,16]
lr = 1e-2
epochs = 300
langevin_steps = 5
step_size = 0.1

model = MLP(n, hidden_sizes)
optimizer = optim.Adam(model.parameters(), lr=lr)

# --------- Dataset: all 2^n vectors ---------
all_vectors = np.array(list(product([0,1], repeat=n)))
X_tensor = torch.tensor(all_vectors, dtype=torch.float32)

# Group indices by k ones
buckets = {k: [] for k in range(n+1)}
for i, vec in enumerate(all_vectors):
    k = vec.sum()
    buckets[k].append(i)

# --------- Training loop with checkpoints ---------
checkpoints = [0, 50, 150, 299]
prob_history = []

for epoch in range(epochs):
    optimizer.zero_grad()
    
    pos = X_tensor
    
    # Negative samples: random init
    neg = torch.rand_like(X_tensor)
    neg.requires_grad = True
    
    # Langevin dynamics
    for _ in range(langevin_steps):
        energy = model(neg).sum()
        grad = torch.autograd.grad(energy, neg, create_graph=False)[0]
        neg.data -= 0.5 * step_size * grad
        neg.data += step_size * torch.randn_like(neg)
        neg.data = torch.clamp(neg.data, 0, 1)

    # Energies
    E_pos = model(pos).mean()
    E_neg = model(neg.detach()).mean()
    
    # Contrastive divergence
    loss = E_pos - E_neg
    loss.backward()
    optimizer.step()
    
    if epoch in checkpoints:
        with torch.no_grad():
            energies = model(X_tensor).squeeze().numpy()
            probs = np.exp(-energies)
            probs /= probs.sum()
            prob_history.append(probs.copy())

# --------- Plotting probabilities by k buckets ---------
labels = [''.join(map(str, v)) for v in all_vectors]

plt.figure(figsize=(15,5))
for i, probs in enumerate(prob_history):
    plt.subplot(1, len(prob_history), i+1)
    
    start = 0
    for k in range(n+1):
        idxs = buckets[k]
        plt.bar(np.array(idxs), probs[idxs], label=f'k={k}')
        if k < n:
            plt.axvline(x=idxs[-1]+0.5, color='k', linestyle='--')
    
    plt.xticks(range(len(probs)), labels, rotation=45)
    plt.ylim(0,1)
    plt.title(f'Epoch {checkpoints[i]}')
    if i == 0:
        plt.ylabel('Probability (EBM)')
    if i == len(prob_history)-1:
        plt.legend()

plt.suptitle('EBM Probabilities for All Binary Vectors by k Ones')
plt.tight_layout()
plt.show()
```

**Comment by josephmaa** [comment]
Not the best gradients of the energy for our puzzles: Robot recommended: Great — thanks for sharing the plot! Let’s walk through what’s happening in each panel:

Top left: Loss (data – neg)

It plunges negative very fast, and keeps drifting down.

Ideally, you want it to decrease then flatten, but here it just keeps going.

That means the model is pushing the energy of negatives way too high relative to data, instead of settling.

Top middle: Energy(data)

Energy assigned to true data keeps increasing steadily.

In a healthy run, this should stabilize at a low-ish value (a “basin” for the data).

Rising energies on data → the model is no longer giving the data low energy → it’s actually forgetting them.

Top right: Energy(neg)

Negatives are being pushed to huge positive energies, way higher than data.

That’s expected to some extent, but here it’s exploding instead of leveling off.

Bottom left: Gap (edata – eneg)

Becomes more and more negative, which means edata is higher and higher above eneg.

In theory, we want the gap to converge toward zero (or a small stable offset).

A large negative drift like this = instability.

Bottom middle: Grad norm

Shoots up to the clip limit (=10) and stays pegged there.

That means the optimizer is constantly hitting gradient clipping → the model is unstable.

Often indicates the learning rate or SGLD step size is too high.

Bottom right: SGLD drift

Around 0.27–0.32 consistently, with some noise.

That’s fine — it suggests the sampler is moving, not stuck.

So sampling itself isn’t the bottleneck; it’s the training dynamics.

Diagnosis

The model is diverging rather than converging.

Signals: exploding energies (both data and neg), gap going the wrong way, grad_norm saturated.

This usually means:

Learning rate is too high, lower LR (e.g. 1e-4 instead of 1e-3).

Or SGLD step size (sgld_step_size) is too big → try halving or reducing by 10×.

Or both together.
